src/commands/distribute/lib/mc-fus-uploader/mc-fus-uploader.ts

```diff
@@ -0,0 +1,777 @@
+      uploadStats.AverageSpeedInMbps +
+      " Mbps.";
+    this.log(completeMessage, {
+      UploadFileSize: this.uploadData.File!.size,
+      UploadSpeed: uploadStats.AverageSpeedInMbps,
+      ElapsedSeconds: uploadStats.TotalTimeInSeconds,
+    });
+
+    this.eventHandlers.onCompleted(uploadStats);
+  }
+
+  enqueueChunks(chunks: number[]) {
+    // if the queue is empty then just add all the chunks
+    if (this.uploadStatus.ChunkQueue.length === 0) {
+      this.uploadStatus.ChunkQueue = chunks;
+      return;
+    }
+
+    // if there something in the queue, don't re-add a chunk. This
+    // can result in more than one thread uploading the same chunk
+    this.uploadStatus.ChunkQueue = this.uploadStatus.ChunkQueue.concat(
+      chunks.filter(function (chunk) {
+        return this.uploadStatus.ChunkQueue.indexOf(chunk) < 0;
+      })
+    );
+  }
+
+  error(errorMessage: string, properties: LogProperties = {}, errorCode?: McFusUploadState) {
+    errorCode = errorCode || McFusUploadState.FatalError;
+    this.setState(errorCode);
+    properties.VerboseMessage = "Error Code: " + errorCode + " - " + errorMessage;
+    this.log(errorMessage, properties, McFusMessageLevel.Error);
+  }
+
+  finishUpload() {
+    // Only verify the upload once at a time.
+    if (this.uploadStatus.State === McFusUploadState.Verifying || this.uploadStatus.State === McFusUploadState.Completed) {
+      return;
+    }
+
+    this.setState(McFusUploadState.Verifying);
+    this.log("Verifying upload on server.");
+    const self = this;
+    this.sendRequest({
+      type: "POST",
+      useAuthentication: true,
+      url:
+        self.uploadBaseUrls.UploadFinished +
+        encodeURIComponent(self.uploadData.AssetId) +
+        "?callback=" +
+        encodeURIComponent(self.uploadData.CallbackUrl),
+      error: function (err: Error) {
+        self.log("Finalize upload failed. Trying to autorecover... " + err.message);
+        self.setState(McFusUploadState.Uploading);
+        self.startUpload();
+      },
+      success: function (response: any) {
+        // it's possible that the health check called complete before this method did.
+        // Log the current status and proceed with response verification.
+        if (self.uploadStatus.State !== McFusUploadState.Verifying) {
+          self.log("Verifying: Upload status has changed, current status: " + self.uploadStatus.State);
+        }
+
+        //if no error then execute callback
+        if (response.error === false && response.state === "Done") {
+          self.log("UploadFinalized. McFus reported the upload as completed. Status message: " + response.message, {
+            location: response.location,
+          });
+
+          // Finally report upload completion.
+          self.completeUpload();
+
+          // Attempt to perform a callback
+          self.invokeCallback(response.raw_location);
+        } else {
+          // if chunks are missing enqueue missing chunks
```

Comment:
Genuinely curious: how can this happen?

Reply:
It's a double verification from the backend, but I am not aware if there's any known scenario on how this can happen.
cc @guperrot @owlstronaut

Reply:
Yeah, this was in the original code we received to modify. In fact, there was a bug where they called finishUpload _first_ when an upload started 🤦  so this would always be hit because there was always missing chunks! It definitely shouldn't be hit much now, but if something does think the upload finished and our verification finds it actually isn't done, this saves us.

Reply:
Thanks for the detailed answer, this matches my understanding of the logic.